### PR TITLE
issue 389 split permutation from sponge functions

### PR DESCRIPTION
**Pull request by mssabr01**
I created a new file within oracle called permutations.rs and moved the permutation-related functions out of poseidon.rs. I also created a quick-start for developing on Windows

### CODE REVIEW

File: oracle/src/lib.rs [hunk 1]
```diff
@@ -2,6 +2,7 @@ pub mod constants;
 pub mod pasta;
 pub mod poseidon;
 pub mod sponge;
+pub mod permutations;
```
Comment (mimoo) [inline oracle/src/lib.rs:5]: should be permutation, singular

File: oracle/src/poseidon.rs [hunk 2]
```diff
@@ -50,84 +51,6 @@ pub struct ArithmeticSponge<F: Field, SC: SpongeConstants> {
...
-            state[i].add_assign(x);
-        }
-        for state_i in state.iter_mut() {
-            *state_i = sbox::<F, SC>(*state_i);
-        }
-        apply_mds_matrix::<F, SC>(params, state);
-    }
-}
-
 pub fn poseidon_block_cipher<F: Field, SC: SpongeConstants>(
```
Comment (mimoo) [inline oracle/src/poseidon.rs:54]: this should be moved as well

File: oracle/src/permutations.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            .enumerate()
+        {
+            state[i].add_assign(x);
+        }
+        for state_i in state.iter_mut() {
+            *state_i = sbox::<F, SC>(*state_i);
+        }
+        apply_mds_matrix::<F, SC>(params, state);
+    }
+}
```
Comment (mimoo) [inline oracle/src/permutations.rs:82]: needs a line after that

File: oracle/src/permutations.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
+
```
Comment (mimoo) [inline oracle/src/permutations.rs:1]: can you add a module-level doc comment here? Something like

```rust
//! The permutation module contains the function implementing the permutation used in Poseidon
```

**Review by mimoo — COMMENTED** [review]
Thanks for the PR! Left you some comments :o

**Comment by mssabr01** [comment]
Thanks for the reviews. What does crate:: mean in the use statements at the top of the file? For example `use crate::constants::SpongeConstants;`. I'm having trouble finding the answer googling things like "crate:: rust use" lol. I know that `::constants` would mean constants is an external crate

**Comment by mimoo** [comment]
all the files in `oracle/src/*.rs` can refer to each other as `crate::<file_name>`. Basically it just means "this crate"

**Comment by mssabr01** [comment]
I found the tests that are run in the CI pipeline and ran them to fix what was wrong. Hopefully it works this time :)

**Comment by mimoo** [comment]
I recommend installing vscode with rust-analyzer btw! and enable auto-format on save :o

File: README.md [hunk 5]
```diff
@@ -87,3 +87,9 @@ The project is organized in the following way:
 * [signer/](https://github.com/o1-labs/proof-systems/tree/master/signer). Interfaces for mina signature schemes.
 * [tools/](https://github.com/o1-labs/proof-systems/tree/master/tools). Various tooling to help us work on kimchi.
 * [utils/](https://github.com/o1-labs/proof-systems/tree/master/utils). Collection of useful functions and traits.
+
+## Windows Development
+Windows development can be done using [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install).
+* Install and open WSL
+* Within WSL, install OCaml using your distro's package manager. For example: `apt install opam`
+* Within WSL, navigate to the project directory and run `cargo test`. If there are no failures then everything is set up correctly.
```
Comment (mimoo) [inline README.md:95]: I think this would be better in a `CONTRIBUTING.md` or something like this.

**Review by mimoo — APPROVED** [review]
Thanks again for the PR!

**Comment by mssabr01** [comment]
Ayyy first accepted pr here 💯 🥇 👍 Thanks for the help. Gonna go pick out another issue

**Pull request merged**